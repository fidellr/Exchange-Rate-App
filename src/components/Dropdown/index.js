import React from 'react';
import cls from 'classnames';
import { isUnique } from '../../utils/helpers/unique';


class Dropdown extends React.Component {
  state = {
    selectedItems: null,
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const { selectedItems } = nextProps;
    if (selectedItems) {
      if (!prevState.selectedItems) {
        return { selectedItems }
      }
      if (selectedItems.length !== prevState.selectedItems.length) {
        return { selectedItems }
      }
    }

    return null;
  }

  render() {
    const { data, onClick } = this.props;
    const { selectedItems } = this.state;

    return (
      <div className="dropdown-container u-boxShadowBottom u-marginHorizontal16 u-borderRadius2 u-borderColorGray u-border1 u-marginBottom16 u-backgroundColorWhite">
        {
          data && data.map((item, index) => (
            <div
              key={`dropdownItem-${item.rate_name}`}
              onClick={() => onClick(item)}
              className={cls('dropdown-item u-cursorPointer u-padding4', {
                'u-backgroundColorGrayBlur': (selectedItems && isUnique(selectedItems, 'rate_name', item.rate_name)) === true,
                'u-borderBottomColorGray': index !== data.length - 1,
              })}
            >
              {item.rate_name}
            </div>
          ))
        }
      </div>
    )
  }

}

export default Dropdown;
