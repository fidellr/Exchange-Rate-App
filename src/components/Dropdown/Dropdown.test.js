import React from 'react';
import { shallow } from 'enzyme';
import Dropdown from './index';

const data = [
  {
    rate_name: 'SGD',
    rate_currency: 11.2213,
  },
  {
    rate_name: 'PLN',
    rate_currency: 22.21134,
  },
  {
    rate_name: 'RUB',
    rate_currency: 33.99823,
  }
];

const selectedItems = data.slice(0, 2);

describe('Dropdown', () => {
  it('should render 3 dropdown items', () => {
    const component = shallow(<Dropdown data={data} selectedItems={selectedItems} />);
    expect(component.children()).toHaveLength(3);
    expect(component).toMatchSnapshot();
  });
});

describe('Dropdown', () => {
  it('should render dropdown item in gray color for the selectedItem', () => {
    const component = shallow(<Dropdown data={data} selectedItems={selectedItems} />);
    const childrens = component.children();
    expect(childrens.filterWhere(c => c.hasClass('u-backgroundColorGrayBlur') === true)).toHaveLength(2);
    expect(component).toMatchSnapshot();
  });
});

describe('Dropdown received props', () => {
  it('should fill selectedItems state with item that passed from props through getDerivedStateFromProps', () => {
    const component = shallow(<Dropdown data={data} />);
    expect(component.state().selectedItems).toBeNull();
    component.setProps({ selectedItems });
    expect(component.state().selectedItems).toHaveLength(2);
  });
});

describe('Dropdown received updated selectedItems from props', () => {
  it('should concat selectedItems state with the updated selectedItems that passed from props through getDerivedStateFromProps', () => {
    const component = shallow(<Dropdown data={data} selectedItems={selectedItems} />);
    expect(component.state().selectedItems).toHaveLength(2);
    const updatedSelectedItems = [...selectedItems, {
      rate_name: 'NZD',
      rate_currency: 12.23113,
    }];
    component.setProps({ selectedItems: updatedSelectedItems });
    expect(component.state().selectedItems).toHaveLength(3);
  });
});