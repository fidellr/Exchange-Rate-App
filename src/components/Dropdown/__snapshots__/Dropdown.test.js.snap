// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render 3 dropdown items 1`] = `
<div
  className="dropdown-container u-boxShadowBottom u-marginHorizontal16 u-borderRadius2 u-borderColorGray u-border1"
>
  <div
    className="dropdown-item u-cursorPointer u-padding4 u-backgroundColorGrayBlur u-borderBottomColorGray"
    key="dropdownItem-SGD"
    onClick={[Function]}
  >
    SGD
  </div>
  <div
    className="dropdown-item u-cursorPointer u-padding4 u-backgroundColorGrayBlur u-borderBottomColorGray"
    key="dropdownItem-PLN"
    onClick={[Function]}
  >
    PLN
  </div>
  <div
    className="dropdown-item u-cursorPointer u-padding4"
    key="dropdownItem-RUB"
    onClick={[Function]}
  >
    RUB
  </div>
</div>
`;

exports[`Dropdown should render dropdown item in gray color for the selectedItem 1`] = `
<div
  className="dropdown-container u-boxShadowBottom u-marginHorizontal16 u-borderRadius2 u-borderColorGray u-border1"
>
  <div
    className="dropdown-item u-cursorPointer u-padding4 u-backgroundColorGrayBlur u-borderBottomColorGray"
    key="dropdownItem-SGD"
    onClick={[Function]}
  >
    SGD
  </div>
  <div
    className="dropdown-item u-cursorPointer u-padding4 u-backgroundColorGrayBlur u-borderBottomColorGray"
    key="dropdownItem-PLN"
    onClick={[Function]}
  >
    PLN
  </div>
  <div
    className="dropdown-item u-cursorPointer u-padding4"
    key="dropdownItem-RUB"
    onClick={[Function]}
  >
    RUB
  </div>
</div>
`;
